File: app/src/main/java/betrip/kitttn/architecturecomponentstest/vm/selected_country/SelectedCountryViewModel.java
package betrip.kitttn.architecturecomponentstest.vm.selected_country;

import android.arch.lifecycle.ViewModel;

import java.net.UnknownHostException;

import betrip.kitttn.architecturecomponentstest.services.CountryDetailsLoader;
import betrip.kitttn.architecturecomponentstest.vm.selected_country.state.ErrorSelectedCountryState;
import betrip.kitttn.architecturecomponentstest.vm.selected_country.state.LoadingSelectedCountryState;
import betrip.kitttn.architecturecomponentstest.vm.selected_country.state.SelectedCountryState;
import betrip.kitttn.architecturecomponentstest.vm.selected_country.state.SuccessSelectedCountryState;
import io.reactivex.Flowable;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.processors.PublishProcessor;
import io.reactivex.schedulers.Schedulers;

/**
 * @author kitttn
 */

public class SelectedCountryViewModel extends ViewModel {
    private PublishProcessor<SelectedCountryState> state = PublishProcessor.create();
    private CompositeDisposable composite = new CompositeDisposable();
    private CountryDetailsLoader loader;

    public SelectedCountryViewModel(CountryDetailsLoader loader) {
        this.loader = loader;
    }

    public void fetchCountryDetails(String countryName) {
        state.onNext(new LoadingSelectedCountryState());

        Disposable d = loader.getCountryDetails(countryName)
                .subscribeOn(Schedulers.io())
                .subscribe(res -> state.onNext(new SuccessSelectedCountryState(res)),
                        err -> state.onNext(createSearchError(err)));

        composite.add(d);
    }

    private ErrorSelectedCountryState createSearchError(Throwable error) {
        if (error instanceof UnknownHostException) {
            return new ErrorSelectedCountryState("No internet connection detected!",
                    ErrorSelectedCountryState.ErrorCode.NO_NETWORK);
        }

        return new ErrorSelectedCountryState(error.getMessage(), ErrorSelectedCountryState.ErrorCode.OTHER);
    }

    public Flowable<SelectedCountryState> getCountryDetails() {
        return state;
    }

    @Override protected void onCleared() {
        composite.clear();
        super.onCleared();
    }
}

File: app/src/main/java/betrip/kitttn/architecturecomponentstest/vm/SelectedCountryViewModel.kt
package betrip.kitttn.architecturecomponentstest.vm

import android.arch.lifecycle.ViewModel
import betrip.kitttn.architecturecomponentstest.model.CountryDetails
import betrip.kitttn.architecturecomponentstest.plusAssign
import betrip.kitttn.architecturecomponentstest.services.CountryDetailsLoader
import io.reactivex.Flowable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.processors.PublishProcessor
import io.reactivex.schedulers.Schedulers

/**
 * @author kitttn
 */

sealed class SelectedCountryState

class CountryLoadingState : SelectedCountryState()

data class CountryLoadSuccessState(val country: CountryDetails) : SelectedCountryState()
data class CountryLoadErrorState(val error: String) : SelectedCountryState()

class SelectedCountryViewModel(private val detailsLoader: CountryDetailsLoader) : ViewModel() {
    private val states = PublishProcessor.create<SelectedCountryState>()
    private val composite = CompositeDisposable()

    fun fetchCountryDetails(countryName: String) {
        states.onNext(CountryLoadingState())
        composite += detailsLoader.getCountryDetails(countryName)
                .subscribeOn(Schedulers.io())
                .subscribe({ states.onNext(CountryLoadSuccessState(it)) },
                        { states.onNext(CountryLoadErrorState(it.localizedMessage)) })
    }

    fun countryDetails(): Flowable<SelectedCountryState> = states

    override fun onCleared() {
        composite.clear()
        super.onCleared()
    }
}